hw/src/trezor.rs

```diff
@@ -0,0 +1,424 @@
+			}
+			"pin_matrix_ack" => {
+				if let (Some(path), Some(msg)) = (device_path, message) {
+					let unlocked = self.pin_matrix_ack(&path, &msg)?;
+					serde_json::to_string(&unlocked).map_err(Error::SerdeError)
+				} else {
+					Err(Error::BadMessageType)
+				}
+			}
+			_ => Err(Error::BadMessageType),
+		}
+	}
+
+	/// Select key derivation path for a known chain.
+	pub fn set_key_path(&mut self, key_path: KeyPath) {
+		self.key_path = key_path;
+	}
+
+	/// List connected wallets. This only returns wallets that are ready to be used.
+	pub fn list_devices(&self) -> Vec<WalletInfo> {
+		self.devices.iter().map(|d| d.info.clone()).collect()
+	}
+
+	/// Get wallet info.
+	pub fn device_info(&self, address: &Address) -> Option<WalletInfo> {
+		self.devices.iter().find(|d| &d.info.address == address).map(|d| d.info.clone())
+	}
+
+	fn open_path<R, F>(&self, f: F) -> Result<R, Error>
+		where F: Fn() -> Result<R, &'static str>
+	{
+		let mut err = Error::KeyNotFound;
+		/// Try to open device a few times.
+		for _ in 0..10 {
+			match f() {
+				Ok(handle) => return Ok(handle),
+				Err(e) => err = From::from(e),
+			}
+			::std::thread::sleep(Duration::from_millis(200));
+		}
+		Err(err)
+	}
+
+	fn pin_matrix_ack(&self, device_path: &str, pin: &str) -> Result<bool, Error> {
+		let usb = self.usb.lock();
+		let device = self.open_path(|| usb.open_path(&device_path))?;
+		let t = MessageType::MessageType_PinMatrixAck;
+		let mut m = PinMatrixAck::new();
+		m.set_pin(pin.to_string());
+		self.send_device_message(&device, &t, &m)?;
+		let (resp_type, _) = self.read_device_response(&device)?;
+		match resp_type {
+			// Getting an Address back means it's unlocked, this is undocumented behavior
+			MessageType::MessageType_EthereumAddress => Ok(true),
+			// Getting anything else means we didn't unlock it
+			_ => Ok(false),
+
+		}
+	}
+
+	fn get_address(&self, device: &hidapi::HidDevice) -> Result<Option<Address>, Error> {
+		let typ = MessageType::MessageType_EthereumGetAddress;
+		let mut message = EthereumGetAddress::new();
+		match self.key_path {
+			KeyPath::Ethereum => message.set_address_n(ETH_DERIVATION_PATH.to_vec()),
+			KeyPath::EthereumClassic => message.set_address_n(ETC_DERIVATION_PATH.to_vec()),
+		}
+		message.set_show_display(false);
+		self.send_device_message(&device, &typ, &message)?;
+
+		let (resp_type, bytes) = self.read_device_response(&device)?;
+		match resp_type {
+			MessageType::MessageType_EthereumAddress => {
+				let response: EthereumAddress = protobuf::core::parse_from_bytes(&bytes)?;
+				Ok(Some(From::from(response.get_address())))
+			}
+			_ => Ok(None),
+		}
+	}
+
+	/// Sign transaction data with wallet managing `address`.
+	pub fn sign_transaction(&self, address: &Address, t_info: &TransactionInfo) -> Result<Signature, Error> {
+		let device = self.devices.iter().find(|d| &d.info.address == address).ok_or(Error::KeyNotFound)?;
+		let usb = self.usb.lock();
```

Comment:
Is the lock required for the entire function? Seems that `HidDevice` does not require the lock, does it?
(same comment applies everywhere where `self.usb.lock()` is done).

Reply:
The HidDevice does require a lock because multiple threads trying to read from the HID interface results in an error (something like "HID Interface is already in use").